DM-39582: Add caching for some butler primitives during deserialization

natelust opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

natelust committed:
Add an object for handling deserialization caches

natelust committed:
Optimize memory and load times on deserialization

Often may butler primitives are deserialized at the same time, and
it is useful for these objects to share references to each other.
This reduces load time and memory usage.

natelust committed:
Convert integer ids to UUID early

Downstream code now depends on refs holding UUIDs. Have the yaml
loader convert old style integer ids to UUIDs early rather than
waiting for downstream cleanups.

natelust committed:
Black and isort changes

natelust committed:
Convert test to use UUID instead of int

natelust committed:
Add release documentation

andy-slac commented on doc/changes/DM-39582.api.md:
```diff
@@ -0,0 +1 @@
+Deprecate reconstituteDimensions argument from Quantum.from_simple
```
I think file name should be `removal`, not `api`, according to README.

Add backticks around `reconstituteDimensions` and `Quantum.from_simple` and period at the end.

andy-slac commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -396,9 +408,18 @@ def from_simple(
+            if datasetType is not None
+            else (x.name if (x := simple.datasetType) is not None else "")
+        )
+        key = (simple.id.int, localName)
```
Should `simple.component` be a part of the key?

natelust replied:
I ran across this too, by construction `simple.datasetType` is a string that will either just be the type itself, or if it is a component be a string of the form `type.component`

andy-slac commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -396,9 +408,18 @@ def from_simple(
         # Minimalist component will just specify component and id and
         # require registry to reconstruct
-        if set(simple.dict(exclude_unset=True, exclude_defaults=True)).issubset({"id", "component"}):
+        if not (simple.datasetType is not None or simple.dataId is not None or simple.run is not None):
```
Could you rewrite this as `simple.datasetType is None and simple.dataId is None and simple.run is None`, I think it makes it easier to read?

natelust replied:
That is not logically the same thing, We only want to run this when they are all `False`. But and is greedy, so False will always gobble up anything.

andy-slac commented on python/lsst/daf/butler/core/datastoreRecordData.py:
```diff
@@ -70,6 +71,11 @@ def direct(
+        key = frozenset(dataset_ids)
+        cache = PersistenceContextVars.serializedDatastoreRecordMapping.get()
+        if cache is not None and (value := cache.get(key)) is not None:
+            return value
```
I would not expect many (or maybe any) cache hits for this. `DatastoreRecordData` is per-quantum structure, I do not think any two quanta can have the same set of input datasets?

andy-slac commented on python/lsst/daf/butler/transfers/_yaml.py:
```diff
@@ -64,6 +64,8 @@
 this version of the code.
 """
 
+_refIntId2UUID = defaultdict[int, uuid.UUID](uuid.uuid4)
```
Is this a workaround for input YAML files that still have integer dataset IDs in them? Should we instead fix those YAML files? It may also be better to generate reproducible UUIDs in that case if you want to keep this map.

natelust replied:
I could not be sure all the places where this might be happening, so I opted to fix it here. Good point on the reproducibility I will go with that.

andy-slac reviewed (approved):
Looks OK, one concern is about cache key for DatasetRef maybe needing `component`. I did not comment on docstring style, I think `ruff` should not take care of that.

natelust committed:
Address formatting/MYPY issues

timj committed:
Add some defensive programming to appease mypy

timj committed:
Check that the datastore record has the right class on read

natelust committed:
Fix mypy with flag comparison change

MyPy seems to narrow types somehow when comparing Enum Flags
directly with equality operators. Compare by value instead.

natelust committed:
Address review feedback
